program/src/processor.rs

```diff
@@ -2311,6 +2313,385 @@ impl Processor {
+            let bids = load_bids_mut(&market, bids_ai)?;
+            let asks = load_asks_mut(&market, asks_ai)?;
+
+            for j in 0..128 {
+                let slot_mask = 1u128 << j;
+                if open_orders.free_slot_bits & slot_mask != 0 {
+                    // means slot is free
+                    continue;
+                }
+                let order_id = open_orders.orders[j];
+
+                if open_orders.is_bid_bits & slot_mask != 0 {
+                    match bids
+                        .find_by_key(order_id)
+                        .and_then(|order| bids.get(order))
+                        .and_then(|node| node.as_leaf())
+                    {
+                        None => continue,
+                        Some(bid) => {
+                            remaining_cancel_size = bid.quantity();
+                            break;
+                        }
+                    };
+                } else {
+                    match asks
+                        .find_by_key(order_id)
+                        .and_then(|order| asks.get(order))
+                        .and_then(|node| node.as_leaf())
+                    {
+                        None => continue,
+                        Some(ask) => {
+                            remaining_cancel_size = ask.quantity();
+                            break;
+                        }
+                    };
+                };
+            }
+        }
+
+        check!(expected_cancel_size > 0, MangoErrorCode::Default)?;
+        check!(remaining_cancel_size > 0, MangoErrorCode::Default)?;
+        check!(new_order.side == cancel_order.side, MangoErrorCode::Default)?;
+
+        // if cancel order has been partially filled, reduce the new order size to accommodate
+        let filled_amount = expected_cancel_size.checked_sub(remaining_cancel_size).unwrap();
+        if filled_amount > 0 {
+            let (_coin_lot_size, pc_lot_size) = {
+                let market = load_market_state(spot_market_ai, dex_prog_ai.key)?;
+                (market.coin_lot_size, market.pc_lot_size)
+            };
+
+            let new_max_qty = new_order.max_coin_qty.get().checked_sub(filled_amount).unwrap();
+            new_order.max_coin_qty = NonZeroU64::new(new_max_qty).unwrap();
+            match cancel_order.side {
+                serum_dex::matching::Side::Bid => {
+                    let new_max_pc_qty_including_fees = pc_lot_size
+                        .checked_mul(new_order.limit_price.get())
+                        .unwrap()
+                        .checked_mul(new_order.max_coin_qty.get())
+                        .unwrap();
+                    new_order.max_native_pc_qty_including_fees =
+                        NonZeroU64::new(new_max_pc_qty_including_fees).unwrap();
+                }
+                serum_dex::matching::Side::Ask => {
+                    new_order.max_native_pc_qty_including_fees =
```

Comment:
if there's ever any use to setting this to something else than `u64::max`, we're overriding the users choice here.